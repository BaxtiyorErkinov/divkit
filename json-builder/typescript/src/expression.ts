const expr = Symbol();

export interface DivExpression {
    [expr]: string;
}

class UnsafeDivExpression implements DivExpression {
    public [expr]: string;

    public constructor(expression: string) {
        this[expr] = expression;
    }

    public toJSON(): string {
        return this[expr];
    }

    public toString(): string {
        return this[expr];
    }
}

export function expression(expression: string): DivExpression {
    return new UnsafeDivExpression(expression);
}
