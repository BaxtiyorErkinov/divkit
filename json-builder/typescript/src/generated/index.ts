// Generated code. Do not modify.

export * from './BooleanVariable';
export * from './ColorVariable';
export * from './Div';
export * from './DivAccessibility';
export * from './DivAction';
export * from './DivAlignmentHorizontal';
export * from './DivAlignmentVertical';
export * from './DivAnimation';
export * from './DivAnimationInterpolator';
export * from './DivAppearanceSetTransition';
export * from './DivAppearanceTransition';
export * from './DivAspect';
export * from './DivBackground';
export * from './DivBase';
export * from './DivBorder';
export * from './DivChangeBoundsTransition';
export * from './DivChangeSetTransition';
export * from './DivChangeTransition';
export * from './DivContainer';
export * from './DivCornersRadius';
export * from './DivCount';
export * from './DivCustom';
export * from './DivData';
export * from './DivDimension';
export * from './DivDownloadCallbacks';
export * from './DivDrawable';
export * from './DivEdgeInsets';
export * from './DivExtension';
export * from './DivFadeTransition';
export * from './DivFixedCount';
export * from './DivFixedSize';
export * from './DivFocus';
export * from './DivFontFamily';
export * from './DivFontWeight';
export * from './DivGallery';
export * from './DivGifImage';
export * from './DivGradientBackground';
export * from './DivGrid';
export * from './DivImage';
export * from './DivImageBackground';
export * from './DivImageScale';
export * from './DivIndicator';
export * from './DivInfinityCount';
export * from './DivInput';
export * from './DivLineStyle';
export * from './DivMatchParentSize';
export * from './DivNeighbourPageSize';
export * from './DivPageSize';
export * from './DivPager';
export * from './DivPagerLayoutMode';
export * from './DivPatch';
export * from './DivPercentageSize';
export * from './DivPivot';
export * from './DivPivotFixed';
export * from './DivPivotPercentage';
export * from './DivPoint';
export * from './DivRoundedRectangleShape';
export * from './DivScaleTransition';
export * from './DivSeparator';
export * from './DivShadow';
export * from './DivShape';
export * from './DivShapeDrawable';
export * from './DivSize';
export * from './DivSizeUnit';
export * from './DivSlideTransition';
export * from './DivSlider';
export * from './DivSolidBackground';
export * from './DivState';
export * from './DivStroke';
export * from './DivTabs';
export * from './DivText';
export * from './DivTooltip';
export * from './DivTransform';
export * from './DivTransitionBase';
export * from './DivTransitionSelector';
export * from './DivTransitionTrigger';
export * from './DivTrigger';
export * from './DivVariable';
export * from './DivVisibility';
export * from './DivVisibilityAction';
export * from './DivWrapContentSize';
export * from './IntegerVariable';
export * from './NumberVariable';
export * from './StringVariable';
export * from './UrlVariable';