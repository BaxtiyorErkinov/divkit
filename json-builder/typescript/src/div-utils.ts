import { DivFixedSize, DivMatchParentSize, DivSizeUnit, DivWrapContentSize } from './generated';

export function fixed(value: number, unit: DivSizeUnit | undefined = undefined): DivFixedSize {
    if (unit === undefined) {
        return new DivFixedSize({ value: value });
    }
    return new DivFixedSize({ unit: unit, value: value });
}

export function matchParent(): DivMatchParentSize {
    return new DivMatchParentSize();
}

export function weighted(weight: number): DivMatchParentSize {
    return new DivMatchParentSize({ weight: weight });
}

export function wrapContent(): DivWrapContentSize {
    return new DivWrapContentSize();
}
