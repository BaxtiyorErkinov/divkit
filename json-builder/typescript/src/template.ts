/* eslint-disable @typescript-eslint/ban-types */

import { Div } from './generated/Div';
import { IDivData } from './generated/DivData';

export interface ITemplates {
    [type: string]: Div;
}

export type Type<U, V extends string = string> = U | TemplatePropertyReference<V, U>;

export class TemplatePropertyReference<V extends string = string, U = {}> {
    public name: V;
    private _value?: U;
    public constructor(name: V) {
        this.name = name;
    }
}

export function reference<V extends string = string, U = {}>(name: V): TemplatePropertyReference<V, U> {
    return new TemplatePropertyReference(name);
}

export class TemplateBlock<T extends string = any, U = object> {
    public readonly type: T;
    public readonly _props?: U;

    public constructor(type: T, props?: U) {
        this.type = type;

        Object.assign(this, props);
    }

    public getProps(): object {
        return Object.keys(this).reduce((acc, k) => {
            if (k !== 'type') {
                acc[k] = true;
            }
            return acc;
        }, {});
    }
}

export function template<T extends string = any, U = object>(type: T, props?: U): TemplateBlock<T, U> {
    return new TemplateBlock(type, props);
}

export function divCard(templates: ITemplates, card: IDivData): { templates: ITemplates; card: IDivData } {
    return {
        templates: templates,
        card: card,
    };
}
