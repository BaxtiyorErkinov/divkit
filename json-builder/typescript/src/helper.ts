import { ITemplates } from './template';

export type NonEmptyArray<T> = T[];

export type Exact<TBase, TExt extends TBase> = TExt extends unknown ? Exactly<TBase, TExt> : never;

type Exactly<TBase, TExt extends TBase> = {
    [K in keyof TExt]: K extends keyof TBase ? TExt[K] : never;
};

export type IntBoolean = 1 | 0;

/**
 * DFS-Обход js-объекта как дерева с выполнением в каждом
 * узле заданного действия
 * @param tree js-объект
 * @param nodeAction действие, выполняемое для каждого узла дерева
 */
export function treeWalkDFS(tree: unknown, nodeAction: (x: unknown, path: string[]) => void): void {
    const stack: [unknown, string[]][] = [[tree, []]];

    while (stack.length) {
        const [node, path] = stack.shift() as [unknown, string[]];
        nodeAction(node, path);

        if (typeof node === 'object' && node !== null) {
            stack.unshift(
                ...Object.entries(node as { [k: string]: unknown }).map(
                    ([key, obj]) => [obj, [...path, key]] as [unknown, string[]],
                ),
            );
        }
    }
}

export function copyTemplates<T extends ITemplates>(templates: T): T {
    const copy: ITemplates = {};

    const placeHolder = (node: unknown): unknown => {
        if (!node || ['string', 'number'].includes(typeof node)) {
            return node;
        }

        if (Array.isArray(node)) {
            return [];
        }

        return Object.create(Object.getPrototypeOf(node));
    };

    const clone = (node: unknown, path: string[]): void => {
        let parentPointer: object = copy;

        path = [...path];
        while (path.length > 1) {
            parentPointer = parentPointer[path.shift() as string];
        }

        const cur = path[0];

        if (!cur) {
            return;
        }

        parentPointer[cur] = placeHolder(node);
    };

    treeWalkDFS(templates, clone);

    return copy as T;
}
