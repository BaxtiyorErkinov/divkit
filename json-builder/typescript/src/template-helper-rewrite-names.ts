import { copyTemplates, treeWalkDFS } from './helper';
import { ITemplates, TemplateBlock } from './template';
import { isExternalTemplate, runResolveDeps } from './template-helper-deps';

/**
 * Переписывание шаблонов с заменой имен вложенных шаблонов
 * Например, в шаблоне
 * new DivContainer({
 *   items: [
 *       template('template2')
 *   ]
 * })
 * 'template2' будет заменен на `template2/${hash(templates.template2)}`
 * @param templates шаблоны для перезаписи
 * @param rename функция, возвращающая новое имя
 * @param depsResolved ранее вычисленные имена, например общих шаблонов
 * @returns переписанные шаблоны + хэшмап новых имен
 */
export function rewriteNames<T extends ITemplates>(
    templates: T,
    rename: (name: string, templates: T) => string,
    resolvedNames: { [k: string]: string } = {},
): {
    templates: T;
    resolvedNames: { [k: string]: string };
} {
    templates = copyTemplates(templates);

    resolvedNames = runResolveDeps(
        templates,
        ({ name, template, depsResolved }) => {
            treeWalkDFS(template, (node) => {
                if (node instanceof TemplateBlock && !isExternalTemplate(node.type)) {
                    (node as any).type = depsResolved[node.type];
                }
            });

            return rename(name, templates);
        },
        resolvedNames,
    );

    return { templates, resolvedNames };
}
