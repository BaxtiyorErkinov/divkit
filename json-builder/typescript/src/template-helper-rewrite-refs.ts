import { copyTemplates } from './helper';
import { ITemplates, TemplatePropertyReference } from './template';

function rewriteRefsTemplate(template: unknown): void {
    const stack: unknown[] = [template];

    while (stack.length > 0) {
        const obj = stack.pop();

        if (obj && typeof obj === 'object') {
            for (const key in obj) {
                const objChild = obj[key];
                if (objChild && typeof objChild === 'object') {
                    if (objChild instanceof TemplatePropertyReference) {
                        obj[key] = undefined; // delete ломает валидатор

                        if (objChild.name !== key) {
                            obj['$' + key] = objChild.name;
                        }
                    } else {
                        stack.push(objChild);
                    }
                }
            }
        }
    }
}

/**
 * Заменяет  в шаблонах замену параметры вида
 * prop: reference('template_prop')
 * на
 * $prop: 'template_prop'
 * @param templates Шаблоны
 * @returns Шаблоны, готовые для сериализации в формате DivKit
 */
export function rewriteRefs<T extends ITemplates>(templates: T): T {
    const result = copyTemplates(templates);

    for (const template of Object.values(result)) {
        rewriteRefsTemplate(template);
    }

    return result;
}
