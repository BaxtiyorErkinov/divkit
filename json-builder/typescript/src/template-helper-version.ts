import { createHash } from 'crypto';
import { Div } from './generated/Div';
import { ITemplates } from './template';
import { rewriteNames } from './template-helper-rewrite-names';

export function getTemplateHash(template: Div): string {
    return createHash('md5').update(JSON.stringify(template), 'utf8').digest('hex');
}

function getHashes(templates: ITemplates): { [k: string]: string } {
    const hashes: { [k: string]: string } = {};

    for (const template of Object.keys(templates)) {
        hashes[template] = getTemplateHash(templates[template]);
    }

    return hashes;
}

function versionName(name: string, hash: string): string {
    return `${name}/${hash}`;
}

export function thelperVersion(templates: ITemplates): (name: string) => string {
    const hashes = getHashes(templates);

    return (name: string) => versionName(name, hashes[name]);
}

/**
 * Переписывает вызовы TemplateBlock в шаблонах,
 * добавляя к их именам версии (md5 хэш).
 * Имена корневых шаблонов (ключи объекта templates) не меняются!
 * @param templates шаблоны
 * @param resolvedNames ранее вычисленные имена, например общих шаблонов
 * @returns хэшмап новых имен шаблонов
 */
export function rewriteTemplateVersions<T extends ITemplates>(
    templates: T,
    resolvedNames?: { [k: string]: string },
): {
    templates: T;
    resolvedNames: { [k: string]: string };
} {
    const hashes: { [k: string]: string | undefined } = {};
    const rename = (name: string, templates: T): string => {
        const template = templates[name];

        if (!template) {
            throw new Error(`No template ${name}`);
        }

        const hash = hashes[name] ?? getTemplateHash(template);

        return versionName(name, hash);
    };

    return rewriteNames(templates, rename, resolvedNames);
}
