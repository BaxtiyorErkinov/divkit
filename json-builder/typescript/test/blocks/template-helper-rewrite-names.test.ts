import { DivContainer, DivText, rewriteNames, template } from '../../src';

describe('Rewrite templates', () => {
    it('should shoud preserve home:block', () => {
        const rename = (name: string): string => `${name}/test`;
        let templates = {
            block1: new DivContainer({
                items: [template('block2'), template('block3')],
            }),
            block2: new DivContainer({
                items: [
                    new DivText({
                        text: 'text1',
                    }),
                ],
            }),
            block3: template('block2'),
            block4: template('home:block'),
        };

        templates = rewriteNames(templates, rename).templates;
        expect(templates).toMatchSnapshot();
    });

    it('should validate template dependencies', () => {
        const templates = {
            template1: template('a', {
                items: [template('template2'), template('b')],
            }),
            template2: template('home:block'),
        };
        expect(() => rewriteNames(templates, (x) => x)).toThrowError(
            `template 'template1' unsolvable dependencies: 'a','b'`,
        );
    });

    it('should catch cyclic dependecies', () => {
        const templates = {
            template1: new DivContainer({
                items: [template('template2')],
            }),
            template2: new DivContainer({
                items: [template('template3')],
            }),
            template3: new DivContainer({
                items: [template('template1')],
            }),
        };
        expect(() => rewriteNames(templates, (x) => x)).toThrowError('cyclic depdendencies');
    });
});
