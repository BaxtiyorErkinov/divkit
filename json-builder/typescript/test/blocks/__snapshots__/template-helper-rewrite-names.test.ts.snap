// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rewrite templates should shoud preserve home:block 1`] = `
{
  "block1": {
    "type": "container",
    "items": [
      {
        "type": "block2/test"
      },
      {
        "type": "block3/test"
      }
    ]
  },
  "block2": {
    "type": "container",
    "items": [
      {
        "type": "text",
        "text": "text1"
      }
    ]
  },
  "block3": {
    "type": "block2/test"
  },
  "block4": {
    "type": "home:block"
  }
}
`;
